### PR TITLE
Tcp broker server

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will add broker servers for zmqss, tcp and udp sockets,  fix some issues with the zmqss socket and add an experimental webserver for use with the broker_server.  

### Proposed changes

- add tcp broker server
- add udp broker server
- add zmqss broker server
- add connection_request and ack for tcp, udp, and zmqss broker servers 
- add disable mechanism for connection request
- add tests of the different broker servers
- add example case for the webserver
- merge all the broker server tests
- fix some issues with the zmqss core type.  ( it is unlikely to be backwards compatible).

### CODE REVIEW

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- src/helics/apps/AsioBrokerServer.cpp  2
- tests/helics/core/TcpCore-tests.cpp  1
- benchmarks/helics/messageSendBenchmarks.cpp  2
- tests/helics/core/ZeromqSSCore-tests.cpp  2
- src/helics/core/tcp/TcpComms.cpp  1
- src/helics/apps/helicsWebServer.cpp  2
         

Clones removed
==============
+ src/helics/apps/helics-broker.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=4928504&bid=16303905)

**Comment by phlptp** [comment]
This PR does not include updated docs.  That will come in a separate PR after this.

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones added
============
- tests/helics/core/ZeromqSSCore-tests.cpp  2
- src/helics/apps/AsioBrokerServer.cpp  2
- src/helics/core/tcp/TcpComms.cpp  1
- src/helics/apps/helicsWebServer.cpp  2
- benchmarks/helics/messageSendBenchmarks.cpp  2
- tests/helics/core/TcpCore-tests.cpp  1
         

Clones removed
==============
+ src/helics/apps/helics-broker.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=4928504&bid=16303905)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/apps/AsioBrokerServer.cpp  2
- tests/helics/core/ZeromqSSCore-tests.cpp  2
- src/helics/core/tcp/TcpComms.cpp  1
- src/helics/apps/helicsWebServer.cpp  2
- tests/helics/core/TcpCore-tests.cpp  1
- benchmarks/helics/messageSendBenchmarks.cpp  2
         

Clones removed
==============
+ src/helics/apps/helics-broker.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=4928504&bid=16303905)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/core/tcp/TcpComms.cpp  1
- tests/helics/core/TcpCore-tests.cpp  1
- benchmarks/helics/messageSendBenchmarks.cpp  2
- src/helics/apps/helicsWebServer.cpp  2
- tests/helics/core/ZeromqSSCore-tests.cpp  2
- src/helics/apps/AsioBrokerServer.cpp  2
         

Clones removed
==============
+ src/helics/apps/helics-broker.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=4928504&bid=16303905)

File: src/helics/apps/CMakeLists.txt [hunk 1]
```diff
@@ -9,6 +9,10 @@
 
 if(HELICS_BUILD_APP_LIBRARY)
 
+    cmake_dependent_option(
+        HELICS_ENABLE_WEBSERVER "Enable a webserver app structure to be embedded with the broker server" OFF
+        "NOT HELICS_DISABLE_BOOST;Boost_VERSION VERSION_GREATER 1.69" OFF
```
Comment (nightlark) [inline src/helics/apps/CMakeLists.txt]: It looks like Boost didn't release any 1.69.x versions, so `Boost_VERSION VERSION_GREATER 1.69` is probably okay.

Comment (phlptp) [reply src/helics/apps/CMakeLists.txt]: I am kind of looking forward to being able to use `VERSION_GREATER_EQUAL` with HELICS 3.0

**Comment by phlptp** [comment]
I have seen a few http libraries that build on top of it.  If you were building a larger application looking into those like [BeastHttp](https://github.com/0xdead4ead/BeastHttp) would probably be wise.  

At some point it will make sense to pull some of that boilerplate into a separate library if we end up using it in multiple locations.    But the fact that it is header only is nice.  I do kind of wish I could make it use the standalone asio, but it is nice that it is header only with the recent versions of boost.

**Comment by nightlark** [comment]
BeastHttp looks better, though I think it's more of the language; the C++ web frameworks I've come across have all felt a bit odd (maybe with the newer C++ features it will get better).

It could be interesting to see how much work it would be to make an Asio version of Beast.

**Pull request merged**